File: tests/mqtt/web/test_mint.py
"""Script to mint a web MQTT connection token.

Flow:
1. Log in with username/password to obtain the auth_session cookie.
2. Call the authenticated user mint endpoint to get a JWT and broker URL.
"""

from __future__ import annotations

import os
import sys
from typing import Tuple

import requests
from dotenv import load_dotenv


# Allow imports from tests/_utils
sys.path.append("tests")
from _utils.jwt_tools import pretty_print_jwt


load_dotenv()

BASE_URL = os.getenv("WEB_APP_BASE_URL", "http://localhost:5173")
USERNAME = os.getenv("SAMPLE_ADMIN_USERNAME", "admin@admin.com")
PASSWORD = os.getenv("SAMPLE_ADMIN_PASSWORD", "admin0823")

LOGIN_URL = f"{BASE_URL}/auth/login?/login"
MINT_URL = f"{BASE_URL}/api/user/mqtt/mint"


def login_and_get_session() -> requests.Session:
    """Log in via the form action and return a session with auth_session cookie set."""

    session = requests.Session()

    response = session.post(
        LOGIN_URL,
        data={"email": USERNAME, "password": PASSWORD},
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        timeout=10,
    )

    print("Login status:", response.status_code)
    try:
        payload = response.json()
        print("Login payload type:", payload.get("type"))
    except Exception:
        print("Login response (non-JSON):", response.text[:200])
        raise

    if response.status_code != 200 or payload.get("type") != "success":
        raise RuntimeError(f"Login failed: status={response.status_code}, body={response.text[:200]}")

    if not session.cookies.get("auth_session"):
        raise RuntimeError("Login succeeded but auth_session cookie was not set")

    return session


def mint_user_mqtt_credentials(session: requests.Session) -> Tuple[str, str, str]:
    """Call the user MQTT mint endpoint using the authenticated session.

    Returns (jwt, brokerUrl, mqttUsername).
    """

    response = session.post(
        MINT_URL,
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        json={},
        timeout=10,
    )

    print("Mint status:", response.status_code)
    payload = response.json()
    print("Raw mint payload:")
    print(payload)

    if not payload.get("success"):
        raise RuntimeError(f"Mint failed: {payload}")

    data = payload.get("data") or {}
    jwt_token = data.get("jwt")
    broker_url = data.get("brokerUrl")
    mqtt_username = data.get("mqttUsername")

    if not jwt_token or not broker_url:
        raise RuntimeError(f"Mint response missing jwt or brokerUrl: {payload}")

    return jwt_token, broker_url, mqtt_username


def main() -> None:
    print("BASE_URL:", BASE_URL)
    print("LOGIN_URL:", LOGIN_URL)
    print("MINT_URL:", MINT_URL)
    print("USERNAME:", USERNAME)

    session = login_and_get_session()
    jwt_token, broker_url, mqtt_username = mint_user_mqtt_credentials(session)

    print("\nBroker URL:", broker_url)
    print("MQTT Username:", mqtt_username)
    print("JWT (first 40 chars):", jwt_token[:40] + "...")

    # Pretty-print decoded JWT header and payload (no signature verification)
    pretty_print_jwt(jwt_token)


if __name__ == "__main__":
    main()

File: tests/mqtt/web/test_mint.ts
import 'dotenv/config';

const BASE_URL = process.env.WEB_APP_BASE_URL ?? 'http://localhost:5173';
const USERNAME = process.env.SAMPLE_USER_USERNAME ?? 'admin@admin.com';
const PASSWORD = process.env.SAMPLE_USER_PASSWORD ?? 'admin0823';

const LOGIN_URL = `${BASE_URL}/auth/login?/login`;
const MINT_URL = `${BASE_URL}/api/user/mqtt/mint`;

console.info(`Minting MQTT credentials for user ${USERNAME}:${PASSWORD}`);

export interface MintResult {
  jwt: string;
  brokerUrl: string;
  mqttUsername?: string;
}

async function loginAndGetCookie(): Promise<string> {
  const body = new URLSearchParams({
    email: USERNAME,
    password: PASSWORD
  });

  const res = await fetch(LOGIN_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body
  });

  let json: any;
  try {
    json = await res.json();
  } catch (err) {
    throw new Error(`Failed to parse login JSON: ${(await res.text()).slice(0, 200)}`);
  }

  if (res.status !== 200 || json?.type !== 'success') {
    throw new Error(`Login failed: status=${res.status}, body=${JSON.stringify(json).slice(0, 200)}`);
  }

  const setCookie = res.headers.get('set-cookie');
  if (!setCookie) {
    throw new Error('Login succeeded but no Set-Cookie header found');
  }

  const match = setCookie.match(/auth_session=([^;]+)/);
  if (!match) {
    throw new Error(`Set-Cookie header does not contain auth_session: ${setCookie}`);
  }

  return `auth_session=${match[1]}`;
}

async function mintUserMqtt(cookie: string): Promise<MintResult> {
  const res = await fetch(MINT_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Cookie: cookie
    },
    body: JSON.stringify({})
  });

  const payload: any = await res.json();

  if (!payload?.success) {
    throw new Error(`Mint failed: ${JSON.stringify(payload)}`);
  }

  const data = payload.data ?? {};
  const jwt = data.jwt as string | undefined;
  const brokerUrl = data.brokerUrl as string | undefined;
//   const mqttUsername = data.mqttUsername as string | undefined;

  if (!jwt || !brokerUrl) {
    throw new Error(`Mint response missing jwt or brokerUrl: ${JSON.stringify(payload)}`);
  }

  return { jwt, brokerUrl };
}

export async function mint(): Promise<MintResult> {
  const cookie = await loginAndGetCookie();
  return mintUserMqtt(cookie);
}

const result = await mint();

console.log(result);

// Pretty-print JWT claims for inspection
try {
  const parts = result.jwt.split('.');
  if (parts.length === 3) {
    const payload = parts[1];
    const json = Buffer.from(payload, 'base64').toString('utf8');
    const claims = JSON.parse(json);
    console.log('JWT claims:', claims);
  } else {
    console.warn('Unexpected JWT format');
  }
} catch (err) {
  console.error('Failed to decode JWT claims:', err);
}